delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/model/resources/ResourceEvent.scala

```diff
@@ -0,0 +1,170 @@
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.context.ContextValue
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.context.JsonLdContext.keywords
+import ch.epfl.bluebrain.nexus.delta.rdf.jsonld.{CompactedJsonLd, ExpandedJsonLd, JsonLdEncoder}
+import ch.epfl.bluebrain.nexus.delta.sdk.model.identities.Identity
+import ch.epfl.bluebrain.nexus.delta.sdk.model.identities.Identity.Subject
+import ch.epfl.bluebrain.nexus.delta.sdk.model.projects.ProjectRef
+import ch.epfl.bluebrain.nexus.delta.sdk.model.{BaseUri, Event, ResourceRef}
+import io.circe.generic.extras.Configuration
+import io.circe.generic.extras.semiauto.deriveConfiguredEncoder
+import io.circe.{Encoder, Json}
+
+import scala.annotation.nowarn
+
+/**
+  * Enumeration of resource event states
+  */
+sealed trait ResourceEvent extends Event {
+
+  /**
+    * @return the resource identifier
+    */
+  def id: Iri
+
+  /**
+    * @return the project where the resource belongs to
+    */
+  def project: ProjectRef
+
+  /**
+    * @return the collection of known resource types
+    */
+  def types: Set[Iri]
+
+}
+
+object ResourceEvent {
+
+  /**
+    * Event representing a resource creation.
+    *
+    * @param id          the resource identifier
+    * @param project     the project where the resource belongs
+    * @param schema      the schema used to constrain the resource
+    * @param types       the collection of known resource types
+    * @param source      the representation of the resource as posted by the subject
+    * @param compacted   the compacted JSON-LD representation of the resource
+    * @param expanded    the expanded JSON-LD representation of the resource
+    * @param rev         the resource revision
+    * @param instant     the instant when this event was created
+    * @param subject     the subject which created this event
+    */
+  final case class ResourceCreated(
+      id: Iri,
+      project: ProjectRef,
+      schema: ResourceRef,
+      types: Set[Iri],
+      source: Json,
+      compacted: CompactedJsonLd,
+      expanded: ExpandedJsonLd,
+      rev: Long,
+      instant: Instant,
+      subject: Subject
+  ) extends ResourceEvent
+
+  /**
+    * Event representing a resource modification.
+    *
+    * @param id          the resource identifier
+    * @param project     the project where the resource belongs
+    * @param types       the collection of known resource types
+    * @param source      the representation of the resource as posted by the subject
+    * @param compacted   the compacted JSON-LD representation of the resource
+    * @param expanded    the expanded JSON-LD representation of the resource
+    * @param rev         the resource revision
+    * @param instant     the instant when this event was created
+    * @param subject     the subject which created this event
+    */
+  final case class ResourceUpdated(
+      id: Iri,
+      project: ProjectRef,
+      types: Set[Iri],
+      source: Json,
```

Comment:
Can it happen that an update doesn't change the source/compacted/expanded ?
If yes, we can make it optional ?

Reply:
No

Reply:
We could issue a rejection if the source doesn't change ? Or we don't care ?

Reply:
That's currently allowed. The source might be the same but the expanded/compacted form might be different, since remember that it can contain a reference to a context which could have been updated meanwhile.

We could have something that checks if source && compacted && expanded are the same in the updated command and in the current state, to reject. Even though again, that's not the current behaviour

Reply:
It would be a nice optimization, but not in the scope of this PR.